### ISSUE TITLE
LibWeb: The road to Discord

### ISSUE THREAD

**Issue by ben-abraham** [SerenityOS/serenity#12618]
Making an Issue to track all things related to Discord on SerenityOS. :yakthulu:
Please feel free to add anything you find (or fix!) to the list. 🐞 

# Needed Features/HTML Libraries
- [x] `localStorage` implementation 47979996e832bdd05440c4a9437920863543441d
- [x] Workers - Sorta (#12391)
- [ ] Workers - Really
- [ ] AudioContext + Audio ([stubbed here](https://github.com/ben-abraham/serenity/tree/audio-stream))
- [ ] History - Needed by the main app
- [ ] StructuredSerialize - Needed by `History`
- [ ] WebSocket fragmented frames (20kb zlib chunk sent from server)
- [ ] Proper <a>  event handling

# Landing Screens

### [Invite Screen](https://discord.gg/serenityos) Checklist:
- [x] Load without crashing
- [x] SVG Background
- [ ] Serenity logo
- [x] `localStorage` implementation 47979996e832bdd05440c4a9437920863543441d
- [ ] zero-sized/incorrectly sized elements
- [ ] `<input>` uses correct font size 
- [x] Username Input 5f54b8dd6ca114e9b867958f0a64b7b190d47a74
- [x] Accept EULA 246c31ccf6cdc21e5959d7e75b73f4d8fb79fcc5
- [x] Continue button responsive
- [ ] Continue button background
- [ ] Captcha? [Example - crashes a lot](https://patrickhlauke.github.io/recaptcha/)
- [ ] New account details
- [ ] Submit
- [ ] Auto-Login?

### [Invite Screen - Already have an account](https://discord.gg/serenityos) Checklist:
- [ ] Load without crashing (sometimes)
- [x] Username & Password Input
- [ ] QR WebSocket login helper (@Lubrsi is looking at)
- [x] User/Pass validation + submission (https://github.com/SerenityOS/serenity/pull/12837)
- [ ] Login flow handled correctly

# WebApp

**These two screens below currently require you simulate a login by hard-coding your discord auth tokens into localStorage. (Steps at bottom of post)**
Also, there are 2 additional resource loader redirects that need to be made for this to advance, they are listed further down.

### [Direct 'Accept Invite' button](https://discord.gg/serenityos)  - **REQUIRES** localStorage fake login:
- [ ] Sends to `discord://` URI handler, Browser can't handle, no auto-redirect

### [Direct to app](https://discord.com/channels/830522505605283862/832931362058403850) Checklist - **REQUIRES** localStorage fake login:
- [ ] Load without crashing
- [ ] Lots of various SVG crashes #12428 
- [ ] WebSocket issues (@Lubrsi's work probably related)
- [ ] Mis-aligned/stacked elements (like <10% chance the app loads this far)

# Modifications

## JS Errors
To help debug JS errors, it's suggested to make the following change to your `VM.h` file, so that you can fully trace errors even if they aren't logged to console.

```diff
diff --git a/Userland/Libraries/LibJS/Runtime/VM.h b/Userland/Libraries/LibJS/Runtime/VM.h
@@ -166,6 +166,9 @@ public:
     template<typename T, typename... Args>
     Completion throw_completion(GlobalObject& global_object, ErrorType type, Args&&... args)
     {
+        dbgln("Throwing JavaScript Exception (with message): {}", String::formatted(type.message(), forward<Args>(args)...));
+        dump_backtrace();
+
         return throw_completion<T>(global_object, String::formatted(type.message(), forward<Args>(args)...));
     }
```

### Resource Loader Patches (url's will change periodically):
- [Everywhere - Firebug spam](https://discord.com/assets/c5f886df05c0c743ae27.js) `try{return window.Firebug.chrome.isInitialized}catch(e){return!1}` => `return !1;`
- ~~[Direct to app- Regex Failure](https://discord.com/assets/32a385a6252512863141.js) `((?=lg)?[vl]k\-?\d{3}) bui| 3\.[-\w; ]{10}lg?-([06cv9]{3,4})` => `((?=lg)[vl]k\-?\d{3}) bui| 3\.[-\w; ]{10}lg?-([06cv9]{3,4})` (second question mark removed)~~ 1c8bb8a3216b1f7e2e3c631d6de4b5cb2c7412ce
- [Direct to app - Weird deep copy]() `Object.freeze(e);Object.getOwnPropertyNames(e).forEach((function(a){var n=e[a];"object"!=typeof n||Object.isFrozen(n)||t(n)}));` =>  (fully removed)

### Follow these steps for each file to patch
1. Download a copy of the file into `Base/res/html/`
2. Do find+replacement specified above (before prettier)
3. `prettier -w <hash>.js` to clean it up
4. Make the following change to your `ResourceLoader.cpp` file, using the hash of the file you are substituting.

```diff
diff --git a/Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp b/Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp
@@ -116,7 +116,20 @@

 void ResourceLoader::load(LoadRequest& request, Function<void(ReadonlyBytes, const HashMap<String, String, CaseInsensitiveStringTraits>& response_headers, Optional<u32> status_code)> success_callback, Function<void(const String&, Optional<u32> status_code)> error_callback)
 {
-    auto& url = request.url();
+    auto url = request.url();
+
+    if(url == "https://discord.com/assets/<hash>.js") {
+        url = "file:///res/html/<hash>.js";
+    }
```

## localStorage direct login
If you want to attempt to load the webapp directly before the login page is working, you can simulate a session by extracting your `localStorage` tokens from your browser, and then directly setting them for `discord.com` like so:
### DO NOT SHARE YOUR TOKENS! EVER!!!!

```diff
diff --git a/Userland/Libraries/LibWeb/DOM/Window.cpp b/Userland/Libraries/LibWeb/DOM/Window.cpp
@@ -445,8 +445,20 @@ RefPtr<HTML::Storage> Window::local_storage()
     // FIXME: Implement according to spec.
 
     static HashMap<Origin, NonnullRefPtr<HTML::Storage>> local_storage_per_origin;
-    return local_storage_per_origin.ensure(associated_document().origin(), [] {
-        return HTML::Storage::create();
+    return local_storage_per_origin.ensure(associated_document().origin(), [this] {
+        auto storage = HTML::Storage::create();
+
+        if(associated_document().origin().host() == "discord.com") {
+            storage->set_item("token", "\"...\"");
+            storage->set_item("tokens", "{\"...\"}");
+            storage->set_item("user_id_cache", "\"...\"");
+            storage->set_item("email_cache", "\"...\"");
+        }
+
+        return storage;
     });
 }
```

**Comment by paulwratt** [comment]
Apart from that code snippet, there is no mention of "Storage" in any of the tick lists.

Have Andreas `localStarage` commits been reflected to `sessionLocalStorage` yet? (that was in essence what he created, just that that context was not appropriate for "the road to Discod").

Also I think the default INPUT element sizes need to be looked at here too, and all CHECKBOX code should be reflected to RADIO as well:

>the default WIDTH x HEIGHT sizes for INPUT, should be based on the W x H recorded from the layout node, which should be based on what the OS provides as that specific INPUT when rendered in its default state (eg on FF the CHECKBOX was 24x24 - from previous video). ATM the CHECKBOX is inheriting WIDTH (at least) from the Previous NODE (PARENT with LABEL and SYBLING when NO PARENT - thats what I observed in the latest video). Also Note: an INPUT type TEXT without size defines, is 20 characters wide (even thought the used font will be variable-width on most platforms - I think this comes from Mosaic & HotJava using mono-spaced fonts as default).

**Comment by ben-abraham** [comment]
> Apart from that code snippet, there is no mention of "Storage" in any of the tick lists.
> 
> Have Andreas `localStarage` commits been reflected to `sessionStorage` yet? (that was in essence what he created, just that that context was not appropriate for "the road to Discord").
>
> Remember that in order to "auto-login" the current `localStroage` needs a "persistant" backing store, but might be tested ATM with 2 windows in Browser (if it can handle 2x Discord web pages?) where one has already set `localStorage`.

Good catch, I had just glossed over this one somehow, despite me being the one who raised it as being needed 🙃 added to the list.

> Also I think the default INPUT element sizes need to be looked at here too, and all CHECKBOX code should be reflected to RADIO as well:
> 
> > the default WIDTH x HEIGHT sizes for INPUT, should be based on the W x H recorded from the layout node, which should be based on what the OS (widget?) provides as that specific INPUT when rendered in its default state (eg on FF the CHECKBOX was 24x24 - from previous video). ATM the CHECKBOX is inheriting WIDTH (at least) from the Previous NODE (PARENT with LABEL and SYBLING when NO PARENT - thats what I observed in the latest video). Also Note: an INPUT type TEXT without size defines, is 20 characters wide (even thought the used font will be variable-width on most platforms - I think this comes from Mosaic & HotJava using mono-spaced fonts as default).

Added items for wrongly-sized items and using the correct font size.

**Comment by paulwratt** [comment]
Glad to help

> Added items for wrongly-sized items and using the correct font size.

That will filter throught to most of the INPUT tags

Specifically for the "correct font size", there may even be something in the spec about INPUT type TEXT by now, but if there is not, what ever the default mono-spaced font x20 (presuming 20 is still the default SIZE attribute) - NOTE: this is _not_ the same as WIDTH & HEIGHT attributes (which reflect pixels), or the STYLE attributes (which also reflect points, EM, etc.)

There is a similar _default size_ for the TEXTAREA element too, but I can't recal it of the top of my head. (20x4 maybe?)

**Comment by konradbartecki** [comment]
WebSocket fragmented frames (20kb zlib chunk sent from server) - https://github.com/SerenityOS/serenity/pull/20925
